juju/api.go

```diff
@@ -556,3 +429,85 @@ func addrsChanged(a, b []string) bool {
 	}
 	return false
 }
+
+var cachedAddressesExistErr = errors.New("cached API endpoints unexpectedly exist")
+
+// ControllerUpdateParams holds controller details for the UpdateControllerAddresses call.
+type ControllerUpdateParams struct {
+	ControllerName string
+	ControllerUUID string
+	CACert         string
+}
+
+// UpdateControllerAddresses writes any new api addresses to the client controller file.
+// Controller may be specified by a UUID, in which case the controller must already exist,
+// or a name, which may be a new or existing controller.
+func UpdateControllerAddresses(
+	store jujuclient.ControllerStore, legacystore configstore.Storage, params ControllerUpdateParams,
+	modelName string, currentHostPorts [][]network.HostPort, addrConnectedTo ...network.HostPort,
+) error {
+	if params.ControllerName == "" && params.ControllerUUID == "" {
+		return errors.New("expected either controller name or UUID")
+	}
+
+	// TODO(wallyworld) - stop storing legacy controller info when
+	// all code ported across to use new yaml files.
+	info, err := legacystore.ReadInfo(modelName)
+	if err != nil {
+		return errors.Annotate(err, "failed to get connection info")
+	}
+	endpoint := info.APIEndpoint()
+
+	var controllerDetails *jujuclient.ControllerDetails
+	if params.ControllerName == "" {
+		// We expect an existing controller.
+		// Look up controller using its uuid.
+		all, err := store.AllControllers()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for name, details := range all {
+			if details.ControllerUUID == params.ControllerUUID {
+				controllerDetails = &details
+				params.ControllerName = name
+				break
+			}
+		}
+	} else {
+		// Look up the controller by name and create it if it doesn't exist.
+		var err error
+		controllerDetails, err = store.ControllerByName(params.ControllerName)
+		if errors.IsNotFound(err) {
+			controllerDetails = &jujuclient.ControllerDetails{
+				ControllerUUID: params.ControllerUUID,
+				CACert:         params.CACert,
+			}
+			endpoint.CACert = params.CACert
+		} else if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	if params.ControllerName == "" {
+		return errors.NotFoundf("controller name with uuid %v", params.ControllerUUID)
+	}
+
+	addrs, hosts, addrsChanged := PrepareEndpointsForCaching(info, currentHostPorts, addrConnectedTo...)
+	if !addrsChanged {
+		// Something's wrong we already have cached addresses?
+		return cachedAddressesExistErr
```

Comment:
There were 2 behaviours originally - one errored and the other didn't. I'll remove the error